### ISSUE TITLE
Ability to select some specific windows to train by mask (BaseWindows)

### ISSUE THREAD

**Issue by ISPritchin** [Nixtla/neuralforecast#904]
### Description

Thanks for a wonderful product. The more I read the source code, the more impressed I am by the quality of this development.

In our problem, the model must be able to predict only some points in the time series that satisfy the conditions. We would like to have a filtering method based on which we could determine whether a given window should be used for training / validation or not.

### Use case

I saw that one of the selection mechanisms is `available_mask`, but its use does not completely solve our problem. Masks can be more specific. May be the library already has a solution for my problem, but I couldn't find it.

I will give an example of a task where window filtering is required. In the simplest case, let us be given vector `y`. But we would like to be able to make a forecast not for all (`'unique_id'`, `'ds'`) points, but for those that satisfy some criteria. I mean we have to select some time periods for prediction, which are specific to every individual client. The criterion can filter out a very large number of windows from training. This criterion can be calculated in advance by the user (calculated before training and provided by the user to the model).

Suggested solution:

- I suggest to enter another reserved name for the column (like `'unique_id'`, `'ds'`,  `'available_mask'`). For the example below I'm using the name `'is_used'`.
- If the last value in the `input_size`-part of this column is 1, we select the window, otherwise not.

Example:
```
input_size = 3
h = 2
y       = [3, 4, 5, 6, 7, 8, 9, 10]
is_used = [0, 0, 1, 0, 1, 1, 0, 0]
```

Currently the following windows would be obtained:
```
[
    [3, 4, 5, 6, 7],
    [4, 5, 6, 7, 8],
    [5, 6, 7, 8, 9],
    [6, 7, 8, 9, 10]
]
```

But using the new column we would like to get:
`[[3, 4, 5, 6, 7], [5, 6, 7, 8, 9], [6, 7, 8, 9, 10]`

The windows were selected because `is_used[input_size - 1] ` is 1
```
y = [3, 4, 5, 6, 7],  is_used = [0, 0, 1, 0, 1]
y = [5, 6, 7, 8, 9],  is_used = [1, 0, 1, 1, 0] 
y = [6, 7, 8, 9, 10], is_used = [0, 1, 1, 0, 0] 
```

I am convinced that the implementation of this functionality will greatly increase the capabilities of the library. I will provide you with any information you need to resolve this issue.

Thanks for your hard work. I do not rule out that someday me and my team will be able to join the contributors to your project.

**Comment by ISPritchin** [comment]
Were you able to understand the idea described above? I, in fact, completed the implementation locally and, it seems, is ready for the pull request.

**Comment by elephaint** [comment]
Thanks for the suggestion, I think I understand the request. Feel free to file a PR with the request so that we can review.

**Comment by mw66** [comment]
> I, in fact, completed the implementation locally and, it seems, is ready for the pull request.

@ISPritchin have you submitted the PR?

This feature is greatly needed:

https://github.com/Nixtla/neuralforecast/issues/1411

**Comment by ISPritchin** [comment]
> > I, in fact, completed the implementation locally and, it seems, is ready for the pull request.
> 
> [@ISPritchin](https://github.com/ISPritchin) have you submitted the PR?
> 
> This feature is greatly needed:
> 
> [#1411](https://github.com/Nixtla/neuralforecast/issues/1411)

@mw66, no, we did not make a pull request for this issue. Our team wrote the solution on a copy of this repository. We have made other changes as well. But due to the lack of experience of interacting with other teams in the implementation of such functionality, we left these changes locally. We think that the development team will be able to make a more elegant solution than ours.

**Comment by mw66** [comment]
> We think that the development team will be able to make a more elegant solution than ours.

I believe that too, the development team know the inner wirings much better. 

This should not be a hard change, but it will greatly improve the usability of the library.

In each batch, just selecting the rows with user mask flags. @elephaint what do you think?

**Comment by ISPritchin** [comment]
@mw66, I know of one case that needs to be handled correctly. In imbalanced problems (when there are few points of interest), it can happen that no window is selected, causing problems with loss calculation.

**Comment by marcopeix** [comment]
I worked on a PR where we add a `sample_weight` reserved column to assign different importance weights to individual timesteps. That way, we can emphasize certain periods of the data, but also completely filter out periods. For example, if in a training window the mean `sample_weight` for the forecast portion is 0, this window doesn't contribute to the loss (so it's ignored). I think it solves this issue.
